src/eredis_cluster.erl

```diff
@@ -175,63 +175,91 @@ query(Command, PoolKey) ->
+        % smart clients should not update internal state
+        % https://redis.io/presentation/Redis_Cluster.pdf
+        {error, <<"ASK ", SlotAddressPort/binary>>} ->
+            try_redirect(SlotAddressPort, Transaction);
+
+        Payload ->
+            Payload
+    end.
 
-        % If the tcp connection is closed (connection timeout), the redis worker
-        % will try to reconnect, thus the connection should be recovered for
-        % the next request. We don't need to refresh the slot mapping in this
-        % case
-        {error, tcp_closed} ->
+handle_transaction_result(Result) ->
+    case Result of
+        % If instance down we wait refresh finish and retry
+        {error, no_connection} ->
+            eredis_cluster_monitor:refresh_mapping(),
             retry;
 
-        % Redis explicitly say our slot mapping is incorrect, we need to refresh
-        % it
-        {error, <<"MOVED ", _/binary>>} ->
-            eredis_cluster_monitor:refresh_mapping(Version),
+        % If any other error we retry
+        {error, Reason} ->
+            log_error(["Redis Cluster Error: ~p", [Reason]]),
             retry;
 
-        Payload ->
-            Payload
+        % Successful transactions and pool_empty error just return
+        Payload -> Payload
     end.
-handle_transaction_result(Result, Version, check_pipeline_result) ->
-    case handle_transaction_result(Result, Version) of
+handle_transaction_result(Result, check_pipeline_result) ->
+    case handle_transaction_result(Result) of
        retry -> retry;
        Payload when is_list(Payload) ->
            Pred = fun({error, <<"MOVED ", _/binary>>}) -> true;
-                    (_) -> false
-                 end,
+                     (_) -> false
+                  end,
            case lists:any(Pred, Payload) of
-               false -> Payload;
                true ->
-                   eredis_cluster_monitor:refresh_mapping(Version),
-                   retry
+                   eredis_cluster_monitor:refresh_mapping(),
+                   retry;
+               false ->
+                   Payload
            end;
        Payload -> Payload
     end.
 
--spec throttle_retries(integer()) -> ok.
-throttle_retries(0) -> ok;
-throttle_retries(_) -> timer:sleep(?REDIS_RETRY_DELAY).
+try_redirect(SlotAddressPort, Transaction) ->
+    [Address, Port] = get_address_port(SlotAddressPort),
+    case eredis_cluster_pool:create(Address, Port) of
+        {ok, Pool} -> eredis_cluster_pool:transaction(Pool, Transaction);
+        _ -> retry
+    end.
+
+get_address_port(SlotAddressPort) ->
+    [_Slot, AddressPort | _] = string:split(binary_to_list(SlotAddressPort), " "),
+    [Address, Port] = string:split(AddressPort, ":"),
+    [Address, list_to_integer(Port)].
+
+
+exponential_backoff(0) -> ok;
+exponential_backoff(Counter) ->
+    Time = trunc(math:pow(?EXPONENTIAL_BACKOFF_BASE, Counter)),
```

Comment:
maybe I'm bit-brushing here, but if we know before-hand how long we should wait for each Counter value, can't we instantiate the list of waiting-times at module level and just fetch it?
This way we can save some floating-point computation time. In elixir this could be like:
```
defmodule EredisCluster do
  @retry_limit 6
  @exponential_backoff_base 5
  @backoff_times Enum.map(0..@retry_limit, &(trunc(:math.pow(@backoff_base, &1))))
  # above is the same as:
  #  @backoff_times Enum.map(0..@retry_limit, fn x -> trunc(:math.pow(@backoff_base, x) end)
  ...

  defp exponential_backoff(counter) do
    Enum.at(@backoff_times, counter)
    |> :timer.sleep()
  end
end
```

Reply:
I don't know a way to do this? :(